torchao/quantization/subclass.py

```diff
@@ -592,3 +596,225 @@ def to_qtensor_components(cls, input_float, groupsize=128, inner_k_tiles=8):
+        self.input_quant_func = input_quant_func
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}(data={self.dequantize()}, shape={self.shape}, "
+            f"device={self.device}, dtype={self.dtype}, input_quant_func={self.input_quant_func}, requires_grad={self.requires_grad})"
+        )
+
+    def dequantize(self, output_dtype=None):
+        if output_dtype is None:
+            output_dtype = torch.float32
+
+        return dequantize_affine(self.int_data, self.block_size, self.scale, self.zero_point, self.int_data.dtype, self.quant_min, self.quant_max, output_dtype=output_dtype)
+
+    def __tensor_flatten__(self):
+        return ["int_data", "scales", "zero_point"], [self.quant_min, self.quant_max, self.block_size, self.shape, self.input_quant_func, self.dtype]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        int_data, scale, zero_point = tensor_data_dict["int_data"], tensor_data_dict["scale"], tensor_data_dict["zero_point"]
+        block_size, quant_min, quant_max, shape, input_quant_func, dtype = tensor_attributes
+        return cls(
+            int_data,
+            scale,
+            zero_point,
+            quant_min,
+            quant_max,
+            block_size,
+            shape if outer_size is None else outer_size,
+            input_quant_func=input_quant_func,
+            dtype=dtype,
+            strides=outer_stride,
+        )
+
+    @classmethod
+    def from_float(
+        cls,
+        input_float,
+        mapping_type,
+        block_size,
+        target_dtype,
+        quant_min = None,
+        quant_max = None,
+        eps = None,
+        scale_dtype = None,
+        zero_point_dtype = None,
+        input_quant_func = None,
+    ):
+        scale, zero_point = choose_qparams_affine(input_float, mapping_type, block_size, target_dtype, quant_min, quant_max, eps, scale_dtype, zero_point_dtype)
+        int_data = quantize_affine(input_float, block_size, scale, zero_point, target_dtype, quant_min, quant_max)
+        return cls(
+            int_data,
+            scale,
+            zero_point,
+            quant_min,
+            quant_max,
+            block_size,
+            input_float.shape,
+            input_quant_func=input_quant_func,
+            dtype=input_float.dtype
+        )
+
+    # __torch_function__ = torch._C._disabled_torch_function_impl
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is torch.nn.functional.linear:
+            input_tensor, weight_qtensor, bias = (
+                args[0],
+                args[1],
+                args[2] if len(args) > 2 else None,
+            )
+            if weight_qtensor.input_quant_func is not None:
+                input_tensor = weight_qtensor.input_quant_func(input_tensor)
```

Comment:
this is for dynamic quant for input

Reply:
This feels quite strange that quantized weight tensor is dictating how other tensors should be quantized.

Actually tensor subclass is fine to capture diffrent quant types but it does not feel right to use for dispatcher.

Reply:
exactly what I was thinking, why connect the two like this?

Reply:
oh this is mainly because we are putting all quantization configuration in the weight tensor, see how it is used in test for more details.

Reply:
Oh, but for the unified AffineQuantizedTensor subclass we don't need to.

This is weight-only:

`linear.weight = to_aqt(linear.weight)`

This is dynamic quant:

`linear.weight = to_aqt(linear.weight)`
*and*
modify forward of `linear` to
```
def forward(input):
    input_qt = to_aqt(input)
    return F.linear(input_qt, self.weight).to(input.dtype)
```
instead of
```
def forward(input):
    return F.linear(input, self.weight)
```

Now, how we inject this additional call of `input_qt = to_aqt(input)` etc. is a different question. There we have so far *also* been using Tensor subclasses. But other other options also include FX, module swap, "model swap", etc.


However, for `AffineQuantizedTensor` we don't need to worry about that. It's a building block that'll then be used by components such as `Int8DynamicallyQuantizedLinearWeight` or perhaps even standalone if the user can stomach the manual dtype conversion.

Reply:
I think putting input_quant_func seems to be easier than the alternatives like module swap, fx, model swap etc., but if there is a clean and widely applicable alternative (maybe hooks?) I think it's fine to switch.